### PR TITLE
Refactor online and Replace Bacon.js with RxJS

### PR DESCRIPTION
**Pull request by MikuroXina**
Separated from #790. `react-rx` is added for use `useObservable`, the hook API for `Observable` in RxJS.

### CODE REVIEW

File: bemuse/src/online/level.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { MappingMode, isMappingMode } from 'bemuse/app/entities/Options'
```
Comment (dtinth) [inline bemuse/src/online/level.ts:1]: Unfortunately this introduces a circular dependency between `bemuse/app` and `bemuse/online`. It is intended for the direct subdirectories in `bemuse/src` to be like an npm package (so-called [packlets](https://www.npmjs.com/package/@rushstack/eslint-plugin-packlets)), so they shouldn’t create a circular dependency between each other.

(I haven’t added a circular dependency detecter to the project yet, so there is currently no automated checks for this, sorry.)

I would recommend one of these:

- Duplicate `MappingMode` into `PlayMode` here.
- Move `MappingMode` and `isMappingMode` into `bemuse/rules`.

**Review by dtinth — CHANGES_REQUESTED** [review]
@MikuroXina Woah many thanks for this! This looks really great 😄 

You have the correct intuition — I intend to remove the `impure` module. It was originally intended to be used for dependency injection in pre-TypeScript-being-mainstream era, but now with TypeScript, it’s better to send the dependencies (e.g. `dispatch`) directly. So that’s for performing that refactor! ✅ 👍 

I have one change regarding dependencies between folders.

By the way, Happy New Year! I hope you have a great year ahead.

File: bemuse/src/online/level.ts (hunk 1 above)
Comment (MikuroXina) [reply bemuse/src/online/level.ts]: I moved it into `bemuse/rules` on cd1225d. Thank you for reviewing.

**Review by dtinth — APPROVED** [review]
Looking great now, thanks! 🙏

**Pull request merged**